### ISSUE TITLE
Partially update a MongoDb embedded document

### ISSUE THREAD

**Issue by localnet** [B4nan/mikro-orm#35]
I'm not sure if this is a bug or if is expected behavior, because I can't find documentation about how to work with embedded documents.

entity/locationModel.ts
```typescript
import { Entity, IEntity, ManyToOne, PrimaryKey, Property } from "mikro-orm";
import { ObjectId } from "mongodb";
import LocationData from "../repositories/locationData";
import { GeometryEntity } from "../types/embedded";
import { BusinessModel } from "./businessModel";

@Entity({ collection: "location", customRepository: () => LocationData })
export class LocationModel {
    @PrimaryKey({ type: "ObjectId" })
    public _id!: ObjectId;

    @Property({ type: "GeometryEntity" })
    public geometry: GeometryEntity;

    @ManyToOne({ type: "BusinessModel", entity: () => BusinessModel })
    public business!: BusinessModel;

    public constructor(geometry: GeometryEntity) {
        this.geometry = geometry;
    }
}

export interface LocationModel extends IEntity<string> { }
```

types/embedded.ts
```typescript
export interface GeometryEntity {
    type: string;
    coordinates: number[];
}
```

repositories/locationData.ts
```typescript
import { EntityRepository } from "mikro-orm";
import { LocationModel } from "../entities/locationModel";
import { CreateLocation, EditLocation } from "../types/inputs/location";

export default class LocationData extends EntityRepository<LocationModel> {
    ...

    public async update(input: EditLocation): Promise<LocationModel | null> {
        const locationModel: LocationModel | null = await this.findOne(input.id);

// tslint:disable-next-line: no-console
        console.log(input);

// tslint:disable-next-line: no-console
        console.log(locationModel);

        if (locationModel) {
            locationModel.assign(input);

// tslint:disable-next-line: no-console
            console.log(locationModel);

            try {
                await this.persist(locationModel);
            } catch (error) {
// tslint:disable-next-line: no-console
                console.log(error);
            }
        }

        return locationModel;
    }
}
```

Output result:
```
[query-logger] db.getCollection("location").find({"_id":"5c9d218960ccc21b905f8069"}).limit(1).toArray(); [took 72 ms]
[Object: null prototype] {
  id: '5c9d218960ccc21b905f8069',
  geometry: [Object: null prototype] { coordinates: [ 25, 25 ] } }
LocationModel {
  _id: 5c9d218960ccc21b905f8069,
  geometry: { type: 'Point', coordinates: [ 50, 50 ] },
  business:
   BusinessModel {
     _id: 5c9d155106dd7215cee1da4c,
     locations:
      Collection {
        owner: [Circular],
        items: [],
        initialized: false,
        dirty: false,
        _populated: false },
     __initialized: false } }
LocationModel {
  _id: 5c9d218960ccc21b905f8069,
  geometry: [Object: null prototype] { coordinates: [ 25, 25 ] },
  business:
   BusinessModel {
     _id: 5c9d155106dd7215cee1da4c,
     locations:
      Collection {
        owner: [Circular],
        items: [],
        initialized: false,
        dirty: false,
        _populated: false },
     __initialized: false } }
{ MongoError: Can't extract geo keys: { _id: ObjectId('5c9d218960ccc21b905f8069'), business: ObjectId('5c9d155106dd7215cee1da4c'), geometry: { coordinates: [ 25, 25 ] } }  unknown GeoJSON type: { coordinates: [ 25, 25 ] }
    at Function.create (/home/alejandro/Proyectos/localeco/backend/node_modules/mongodb-core/lib/error.js:43:12)
    at toError (/home/alejandro/Proyectos/localeco/backend/node_modules/mongodb/lib/utils.js:149:22)
    at coll.s.topology.update (/home/alejandro/Proyectos/localeco/backend/node_modules/mongodb/lib/operations/collection_ops.js:1465:39)
    at handler (/home/alejandro/Proyectos/localeco/backend/node_modules/mongodb-core/lib/topologies/replset.js:1155:22)
    at /home/alejandro/Proyectos/localeco/backend/node_modules/mongodb-core/lib/connection/pool.js:397:18
    at process._tickCallback (internal/process/next_tick.js:61:11)
  driver: true,
  name: 'MongoError',
  index: 0,
  code: 16755,
  errmsg:
   'Can\'t extract geo keys: { _id: ObjectId(\'5c9d218960ccc21b905f8069\'), business: ObjectId(\'5c9d155106dd7215cee1da4c\'), geometry: { coordinates: [ 25, 25 ] } }  unknown GeoJSON type: { coordinates: [ 25, 25 ] }',
  [Symbol(mongoErrorContextSymbol)]: {} }
```

It seems that embedded document is overwritten instead of be merged, so my better workaround was:
```typescript
    ...
    const locationModel: LocationModel | null = await this.findOne(input.id);

    if (locationModel) {
        const { geometry, ...location } = input;
        locationModel.assign(location);
        Object.assign(locationModel.geometry, geometry);
        await this.persist(locationModel);
    }
    ...
```

I think there should have a better approach. Some guidance about how to work with MongoDB embedded documents?

**Comment by B4nan** [comment]
Yes, currently it is overridden, not merged. It basically behaves the same way as `Object.assign(entity, data)`, plus it handles converting ID to entity reference. 

I am not sure if deep merging object values should be the default, but we could have `entity.assign()` as it is now, and add `entity.merge()` that would do deep merging, or add new boolean parameter to `entity.assign()`, although that is not self-explanatory. Or maybe `assignDeep` or `assignRecursive`?

Maybe the best approach would be to have the second parameter of `assign()` as options object `{ onlyProperties: boolean, mergeObjects: boolean }`. WDYT?

**Comment by B4nan** [comment]
In the meanwhile, you could implement custom `assign` method on your `LocationModel` entity:

```
import { EntityAssigner } from 'mikro-orm';

@Entity({ collection: "location", customRepository: () => LocationData })
export class LocationModel {

  @PrimaryKey({ type: "ObjectId" })
  public _id!: ObjectId;

  @Property({ type: "GeometryEntity" })
  public geometry: GeometryEntity;

  @ManyToOne({ type: "BusinessModel", entity: () => BusinessModel })
  public business!: BusinessModel;

  public constructor(geometry: GeometryEntity) {
    this.geometry = geometry;
  }
  
  assign(data, onlyProperties = false) {
    const { geometry, ...location } = data;
    EntityAssigner.assign(location, onlyProperties);
    Object.assign(this.geometry, geometry);
  }
  
}
```

**Comment by localnet** [comment]
I think I would prefer add new boolean parameter to `entity.assign()` or to have the second parameter of `entity.assign()` as options object.

Would be possible define the embedded documents as a class and decorate properties with `@Property`, to get property validation, something like this:
```typescript
import { Property } from "mikro-orm";

export class GeometryEntity {
    @Property({ type: "string" })
    public type!: string;

    @Property({ type: "number[]" })
    public coordinates!: number[];
}
```

I am few experience with MongoDB, I'm not sure if normally it will be desirable to have always same default behavior "assign" or "merge" for a embedded document, if so, maybe a new decorator?
```typescript
import { Property } from "mikro-orm";

@Embedded({ merge: true })
export class GeometryEntity {
    @Property({ type: "string" })
    public type!: string;

    @Property({ type: "number[]" })
    public coordinates!: number[];
}
```

Or this can be a lot of mess and diverge too much from SQL.

**Comment by B4nan** [comment]
I will go with the options parameter (while staying BC with the boolean `onlyProperties`). 

The approach with `GeometryEntity` wouldn't currently work out of the box (you would have to handle converting the value to class manually), although that is another possible feature to implement, as well as adding support for embeddable entities. But I don't think one would always need the same `assign/merge` behaviour. 

Let's not over-engineer this, for now the options parameter will be enough :]

**Comment by B4nan** [comment]
Published on npm as v2.1.1. 

https://b4nan.github.io/mikro-orm/entity-helper/
